### ISSUE TITLE
`MinAge` for NodeClaims

### ISSUE THREAD

**Issue by jonathan-innis** [kubernetes-sigs/karpenter#1030]
### Description

**What problem are you trying to solve?**

We are having some discussion around how to control deprovisioning here: https://github.com/aws/karpenter-provider-aws/issues/1738. One mechanism for adding another layer of control is to allow users to set a `minNodeLifetime` or `minAge` on the NodeClaim spec which would indicate to Karpenter that it can't begin to disrupt the node until after it is past its node lifetime. 

This has been mentioned in a few places across a few different issues (like #1014 and #752) but hasn't been formalized into its own separate issue. This effectively allows users to say that they wouldn't like Karpenter to begin disrupting the nodes until after a fixed amount of time. The benefit of this is that Karpenter will not quickly launch and deprovision nodes (even if there is some cost savings) since that churn on the application often means that the application never gets to run.

The use-cases for this has some overlap with #752, #795, and #735

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by Bryce-Soghigian** [comment]
would minAge be specifed on the nodepool in the disruption controls? and be on the nodeclaim itself? This seems to suggest we would put it on the nodeclaim, but it would be rather cumbrous to manage. 

I can see value in the per nodeclaim controls but also would like to manage this through the nodepool.

**Comment by thomaspeitz** [comment]
`would minAge be specifed on the nodepool in the disruption controls?` +1

For us karpenter scales just to fast, and even with budgets it is hard to control.
I want it to scale every few hours and not every few minutes.

This feature would solve it for me and would be a really nice one!

**Comment by Bryce-Soghigian** [comment]
@thomaspeitz  based on your experience what do you think a good default value for minAge should be?

**Comment by Bryce-Soghigian** [comment]
Is anyone already tackling this or can I attempt a design for this?

**Comment by sftim** [comment]
When we come to name the field, try to pick a name that helps people work out the behavior in different scenarios just from the field name:
- I create a NodePool and I specify that I want nodes to last at least an hour. I run some pods and end up with three healthy nodes.
- After 20 minutes, I delete one NodeClaim using `kubectl`. How long does the claimed node survive for?
- After a further 30 minutes, I have three healthy nodes, one only half an hour old. I delete the NodePool. How long does the NodePool take to delete?

**Comment by thomaspeitz** [comment]
So what I did to tune the minAge (kind of)
Going from no budgets (old release)
<img width="344" alt="image" src="https://github.com/kubernetes-sigs/karpenter/assets/8327755/5304c379-f5c3-485e-9b23-2b0db2a1287e">

In detail you can see how bad it is:
<img width="378" alt="image" src="https://github.com/kubernetes-sigs/karpenter/assets/8327755/7e114dcb-e10f-42e3-a9db-7b16de8648d5">

to
this config (which only scales down in the first 10 minutes of an hour)
```
          budgets:
          - nodes: "1"
            schedule: "0 * * * *"
            duration: 10m
          - nodes: "0"
            schedule: "10 * * * *"
            duration: 50m
```


This reduced the (wrong) scaling down a lot.
<img width="235" alt="image" src="https://github.com/kubernetes-sigs/karpenter/assets/8327755/8b33ca7e-6d5f-419c-bdd7-76ca13c532c4">

Blue line cpu allocatable / green line requested cores.

[--scale-down-delay-after-add](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/main.go#L117)
This helps to reduce the scale down.


I would suggest to allow configuring a scale down delay after add. This would solve a lot of problems from my point of view. I could just block scale down for 15m. This would mean less uneeded interruption. In combination of minAge of a node e.g. 15m. This would ensure:
- Nodes are rotated only if they were running for a bit (less costs of image pull on the node) -> Less traffic
- Nodes are scaled down only if they are not needed for x amount of time.

The graphs is the production setup of a client of mine which has a lot of HPAs running.

In the end I won't to avoid that nodes get scaled down, just because some HPA scales down for a short amount of time.
The scaling is to fast and then creates unneeded pod rotations. Devs are unhappy as there pods are restarting all the time.

**Comment by jonathan-innis** [comment]
One thing that we should consider here with the design is the overlap with #735. For you @thomaspeitz, I'm suspecting that either this issue or that issue could help solve your problem. The question that I want to answer here is: What use-cases does `minAge` address that `consolidateAfter` doesn't. Really, `consolidateAfter` says, "wait for nodes to be in an under-utilized state for this long before spinning them down" while `minAge` says, "don't deprovision me for _any_ disruption reason after you have initially spun me up". Both are valid, but I question the utility of blocking drift on a node after it has initially been launched, particularly if its configuration is no longer valid. 

Does `consolidateAfter` just cover everything?

**Comment by jonathan-innis** [comment]
> That being said additional granularity comes at the cost of the configuration being more complex for the end user. I would love to build a DisruptionDefaults/ClusterDefaults CRD so that the end user doesn't have to configure the same thing all throughout their nodepool crs

I'm personally less of a fan of cluster-wide defaults in a separate CRD; since, I think that any default that you want to set on an individual resource could just be managed by your IAC, Kustomize, Helm, or GitOps tooling. @Bryce-Soghigian Have you heard from users that managing the defaults across resources is cumbersome to do with this tooling? 

We've tended to index towards taking away global configuration (AWS provider removed global tagging, in favor of NodeClass-based tagging) exactly for the reason of: if you have the granularity, then you just need a templating engine to manage that consistency across resources.

**Comment by Bryce-Soghigian** [comment]
> but I question the utility of blocking drift on a node after it has initially been launched, particularly if its configuration is no longer valid.

Someone just opened an issue requesting what I believe to be this type of functionality(If I am understanding their usecase correctly) [see here](https://github.com/kubernetes-sigs/karpenter/issues/1047)

`DisruptAfter` or `MinAge` allow blocking all churn or disruption for a specified amount of time.  Some users value this SLO guarantee that you will get this node for 1 week and nothing in the system with automatically remove it.

**Comment by Bryce-Soghigian** [comment]
> Have you heard from users that managing the defaults across resources is cumbersome to do with this tooling? 

You are right of course that you can solve the problem with IAC solutions. But I would like karpenter to manage the configuration of itself and not outsource that responsibility to another solution. 

I have also heard the customer narrative when watching videos on karpenter describing as harder to configure and understand than cluster autoscaler. This begs the question, Why?  CAS has 100 Flags what makes it easier to understand? 

Harder to understand makes sense since we are working with more dimensions via pod + node level controls, node replacements + disruptions etc. There are pros to making this bit more complex. 

With CAS to configure it you go to a single place, the flags, and modify all changes to autoscaling behavior through this one interface. Karpenter has `3 CRDs` for managing this control, plus the helm/deployment configuration values. 

Adding a fourth global configuration CRD won't make the configuration process easier but I would like us to think about ways we can make karpenter simpler, and have observability so they can understand why it makes the decisions it does.

**Comment by edas-smith** [comment]
+1 for this feature

**Comment by Bryce-Soghigian** [comment]
>  would love a link to the videos just to hear directly the reasoning around why people think that we are more confusing 
Let me see if I can dig up some of those videos.

**Comment by Bryce-Soghigian** [comment]
lifecycle frozen
